cpp/include/cudf/fixed_point/floating_conversion.hpp

```diff
@@ -0,0 +1,269 @@
+      // Because of the understood set-bit not stored in the mantissa, it is not possible
+      // to store the value zero directly. Instead both +/-0 and denormals are represented with
+      // the exponent bits set to zero.
+      // Thus it's fastest to just floor (generally unwanted) denormals to zero.
+      return INT_MIN;
+    } else if (exponent_bits == exponent_mask) {
+      //+/-inf and NaN values are stored with all of the exponent bits set.
+      // As none of these are representable by integers, we'll return the same value for all cases.
+      return INT_MIN;
+    }
+
+    // Extract the exponent value: shift the bits down and subtract the bias.
+    using SignedIntegralType                       = cuda::std::make_signed_t<IntegralType>;
+    SignedIntegralType const shifted_exponent_bits = exponent_bits >> num_mantissa_bits;
+    return shifted_exponent_bits - static_cast<SignedIntegralType>(exponent_bias);
+  }
+
+  /** @brief Sets the sign bit of a floating-point number
+   *
+   * @param floating The floating value to set the sign of
+   * @param is_negative The sign bit to set for the floating-point number
+   * @return The input floating-point value with the chosen sign
+   */
+  CUDF_HOST_DEVICE inline static FloatingType set_is_negative(FloatingType floating,
+                                                              bool is_negative)
+  {
+    // Convert floating to integer
+    IntegralType integer_rep = bit_cast_to_integer(floating);
+
+    // Set the sign bit
+    integer_rep |= (IntegralType(is_negative) << sign_bit_index);
+
+    // Convert back to float
+    return bit_cast_to_floating(integer_rep);
+  }
+
+  /** @brief Adds to the base-2 exponent of a floating-point number
+   *
+   * @param floating The floating value to add to the exponent of
+   * @param exp2 The power-of-2 to add to the floating-point number
+   * @return The input floating-point value * 2^exp2
+   */
+  CUDF_HOST_DEVICE inline static FloatingType add_exp2(FloatingType floating, int exp2)
+  {
+    // Convert floating to integer
+    auto integer_rep = bit_cast_to_integer(floating);
+
+    // Extract the currently stored (biased) exponent
+    auto exponent_bits = integer_rep & exponent_mask;
+    auto stored_exp2   = exponent_bits >> num_mantissa_bits;
+
+    // Add the additional power-of-2
+    stored_exp2 += exp2;
+    exponent_bits = stored_exp2 << num_mantissa_bits;
+
+    // Clear existing exponent bits and set new ones
+    integer_rep &= (~exponent_mask);
+    integer_rep |= exponent_bits;
+
+    // Convert back to float
+    return bit_cast_to_floating(integer_rep);
+  }
+};
+
+/** @brief Determine the number of significant bits in an integer
+ *
+ * @tparam T Type of input integer value
+ * @param value The integer whose bits are being counted
+ * @return The number of significant bits: the # of bits - # of leading zeroes
+ */
+template <typename T, typename cuda::std::enable_if_t<(cuda::std::is_unsigned_v<T>)>* = nullptr>
```

Comment:
I believe the trait `is_unsigned_v` will return true for types like `uint16_t`. However, only 32/64/128 bit values appear to be handled below. Should we be more specific in this trait, and only accept 32/64/128-bit unsigned integers? (Will we get a compile error currently for unsupported types?)

Reply:
There's a static assert at the start of the function call so it won't compile. I'll add a @note that only 32/64/128 bit values are allowed.

Reply:
Done.

Reply:
Using both SFINAE and a static_assert seems unnecessary. Can you just make this plain SFINAE?

Reply:
Done